### PR TITLE
(Un)packers for hadronic showers at the uGMT input and output

### PR DESCRIPTION
**Pull request by dinyar**
#### PR description:

This PR extends the `RegionalMuon` and `Muon` unpackers to be able to also unpack `RegionalMuonShower` and `MuonShower` objects respectively. This functionality is gated by the expected uGMT firmware version read from the readout payload.

I've also modified the `RegionalMuonShower` interface to more closely mirror that of the `RegionalMuonCand` data format.

#### PR validation:

Ran `scram build code-checks`, `runTheMatrix.py -l limited -i all --ibeos`, and `scram b runtests` locally.

attn @rekovic, @kbunkow, @dildick

### CODE REVIEW

**Comment by dinyar** [comment]
> I'm afraid I don't know much about data repacking...

Right, however I was curious if the L1TMuonEndCapShowerProducer needed the CSC primitives to be re-emulated? Currently this doesn't seem to be the case.

**Comment by dinyar** [comment]
ping @dildick

**Comment by dildick** [comment]
The showers code seems to be always on, even in Run-1/Run-2. It can probably be hidden behind a flag that enables it only for Run-3, but once you set ```keepShowers``` to True, you be receiving the showers at the EMTF. The EMTF produces showers that feed into the uGMT.

**Comment by dildick** [comment]
Hi @dinyar Thanks for your question. Yes, to run the CSC trigger in a Run-3 era on Run-2 data you have to disable the GEM-CSC trigger since GEM data is not available. You can set ```runME11ILT``` (for Integrated Local Trigger) to ```False```. That should stop the producer from consuming a ```GEMPadDigiClusterCollection```

**Comment by rekovic** [comment]
hold

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
please test

**Comment by jfernan2** [comment]
Hi @dinyar 
I understand that changes are expected from these unpackers but in some cases some stats are lost[1] and in others many changes are seen outside the L1T* folders, is this expected?

Full comparison here: [3]

Thanks

[1] https://tinyurl.com/yxex2onr

[2] https://tinyurl.com/y6zl68vx

[3] https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_12_2_X_2021-11-30-2300+5bb56a/47169/dqm-histo-comparison-summary.html

**Comment by dinyar** [comment]
Hi @jfernan2,

No, I strongly suspect I introduced a bug somewhere. Will have a look. Thanks for the ping!

Cheers,
Dinyar

**Comment by jfernan2** [comment]
please test

**Comment by dinyar** [comment]
Sigh. I did fix the problem in the input packer, but forgot the output packer. Updated commits incoming.. Sorry about that.

**Comment by dinyar** [comment]
Hi @jfernan2,

Could you retest please?

Cheers,
Dinyar

**Comment by perrotta** [comment]
please test

**Comment by dinyar** [comment]
Unless I'm overlooking something I think this looks fine. I don't expect changes to the DQM plots because here we just add (un)packing of showers which aren't yet shown in the DQM.

**Comment by jfernan2** [comment]
+1

**Comment by rekovic** [comment]
unhold

**Comment by rekovic** [comment]
+1

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/GMTCollections.cc [hunk 1]
```diff
@@ -19,6 +19,12 @@ namespace l1t {
       event_.put(std::move(imdMuonsEMTFPos_), "imdMuonsEMTFPos");
       event_.put(std::move(imdMuonsOMTFNeg_), "imdMuonsOMTFNeg");
       event_.put(std::move(imdMuonsOMTFPos_), "imdMuonsOMTFPos");
+
+      event_.put(std::move(regionalMuonShowersEMTF_), "EMTF");
+      event_.put(std::move(muonShowers_[0]), "MuonShower");
+      for (int i = 1; i < 6; ++i) {
```
Comment (perrotta) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/GMTCollections.cc:25]: Could this "6" be made a named constant (if it cannot be taken from somewhere else) to be used also in https://github.com/cms-sw/cmssw/pull/35627/files#diff-b58cb5d2c0f29e48000d3f6f50516977599983b3142acf4599330e59d0268067R74 ?

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/GMTSetup.cc [hunk 2]
```diff
@@ -76,6 +81,12 @@ namespace l1t {
       prod.produces<MuonBxCollection>("imdMuonsEMTFPos");
       prod.produces<MuonBxCollection>("imdMuonsOMTFNeg");
       prod.produces<MuonBxCollection>("imdMuonsOMTFPos");
+
+      prod.produces<RegionalMuonShowerBxCollection>("EMTF");
+      prod.produces<MuonShowerBxCollection>("MuonShower");
+      for (int i = 1; i < 6; ++i) {
```
Comment (perrotta) [inline EventFilter/L1TRawToDigi/plugins/implementations_stage2/GMTSetup.cc:87]: Again this "6" hardcoded...

**Review by perrotta — COMMENTED** [review]
Is this number of muonShowers equal to six fixed by some basic principle? Or could it get modified in some future versions of the algo? 
In general, a named constant made accessible to whoever need to know that size can make the code safer

File: EventFilter/L1TRawToDigi/plugins/implementations_stage2/GMTCollections.cc (hunk 1 above)
Comment (dinyar) [reply EventFilter/L1TRawToDigi/plugins/implementations_stage2/GMTCollections.cc]: Hi Andrea,

True, fair point. I've replaced it by a constant. It's the number of uGT boards we have installed so it's unlikely to be changed however it's now a bit more explicit.

Cheers,
Dinyar

**Comment by perrotta** [comment]
please test

**Comment by jfernan2** [comment]
+1

**Comment by rekovic** [comment]
+1

**Comment by perrotta** [comment]
+1

**Comment by VinInn** [comment]
this breaks  #36396

**Comment by dinyar** [comment]
I suspect 9960aaa7d3e56cdd781986d94b8b7aa14a75c886 caught the other PR out. If that's the issue I can commit a fix for it in the next hour or two I'd say.

**Comment by perrotta** [comment]
As far as I can see the inconsistency is with #36018, which was merged 2 days ago after this PR was tested last
@dinyar if you have a quick fix at hand, please commit the PR: otherwise we'll have to revert this one.

**Comment by VinInn** [comment]
the line that does not compile is
https://github.com/cms-sw/cmssw/pull/36396/files#diff-881ff4a5b7f865747240faa761540efae46b58c1007078bcda43fbfe6f250f6bR62
and another farther below

**Comment by perrotta** [comment]
Also https://github.com/cms-sw/cmssw/pull/36018/files#diff-a2194e73035b8a64c72a7363af81ce438a545aca33b67ce5f616ff6ed98b4d0cR15 and https://github.com/cms-sw/cmssw/pull/36018/files#diff-a2194e73035b8a64c72a7363af81ce438a545aca33b67ce5f616ff6ed98b4d0cR16 , the issue being the removal here of the two methods 
- `RegionalMuonShower::endcap()`
- `RegionalMuonShower::sector()`
which should be restored.

**Comment by zuoxunwu** [comment]
> the line that does not compile is
> https://github.com/cms-sw/cmssw/pull/36396/files#diff-881ff4a5b7f865747240faa761540efae46b58c1007078bcda43fbfe6f250f6bR62
> and another farther below

As @dinyar is already fixing the conflict by encapsulating `RegionalMuonShower::endcap()` and `RegionalMuonShower::sector()`. I should keep these lines as they are, right?

**Comment by VinInn** [comment]
This is what I suggested as the quickest and robust fix to be merged before next IB.
A cleaner solution can be provided at a later time (if deemed necessary)

**Comment by zuoxunwu** [comment]
Agree. Thanks!

**Pull request merged**